Bug 1223124 - Top Site Deletion Rework

sleroux opened the pull request.

thebnich commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -136,16 +138,18 @@ class TopSitesPanel: UIViewController {
+        func reloadThumbnails() -> Success {
+            return self.profile.history.getTopSitesWithLimit(self.layout.thumbnailCount)
+                .bindQueue(dispatch_get_main_queue()) { result in
                     self.deleteOrUpdateSites(result, indexPath: indexPath)
```
It looks like the problem is caused by asynchronously updating the UI based on the index after doing the requery. The index is fragile since any requery during the async update can cause the indices to change.

Rather than doing the query and then updating the UI, what if we delete the item immediately from the collection (in `didRemoveThumbnail`), then fill in the gaps after the query is done? Theoretically, that means we wouldn't need to pass around `indexPath` anymore, which can go stale as we bounce between threads.

sleroux replied:
Your approach makes a lot more sense and simplifies the logic. One issue though is the way we're querying for Top Sites. We're requesting `maxFrecencyLimit` number of cells from the cache to display. `maxFrecencyLimit` is an approximation for the max number of items we'll see in either landscape or portrait orientation. The code currently 'blocks' on the delete until a request is made to get the new set of data from the cache before proceeding. If the user deletes really fast it's easy to get past the number of items we pull from the cache. 

It's similar to the problem of having an infinite list since we don't know how many items there are in total for the Top Sites panel. I mean, we could technically grab the whole cache but since it's calculated based on history it could be huge. Actually, the more I think about it, the more this maps to a producer/consumer problem. The user is 'consuming' items from the panel while the cache queries are 'producing' more items for the panel to display. 

Any thoughts on solving that pickle?

thebnich replied:
Would it be possible to do the deletion and replacement separately? That is:
 1. Immediately delete the item from the collection.
 2. Make queries to delete the item and get the updated data.
 3. Fill in top sites with new thumbnails.

Right now, I know that 1 and 3 happen in the same operation as part of `performBatchUpdates`, which I think we do because of animations. Is there any way we can split them, doing the deletion before any querying, or does delete/replace need to be atomic?

sleroux replied:
Hmm interesting. We can always do two separate performBatchUpdate calls - one for the deletion and another for the replacement. The only requirement is that the data source be in sync before/after each one. I'm going to give that a try. 

Also for https://bugzilla.mozilla.org/show_bug.cgi?id=1226201 I'd like to refactor the panel code to make it more testable so I can add a unit test for that bug so I'll probably do that as part of this work.

sleroux commented:
WIP - Need to clean it up but the general approach I'm going for is:

1. Instead of TopSitesDataSource using the read-only cursor from the DB query, an array is used instead to allow us to insert/delete items from the data source separately from the queries being made to the database.

2. Point 1 allows us to separate the deletion and insertion operations on the collection view to happen asynchronously. didRemoveThumbnail now deletes the item from the data source and data array away, makes the new query, then inserts the next top site cells in place of the deleted ones.

3. The separation in Point 2 helped clean up the methods I had before for deleting items so deletion helper methods are now removed and moved into didRemoveThumbnails.

Still working out some bugs with fewer than numberOfThumbnails items and a bug where duplicate sites are shown.

sleroux commented:
I've pushed up a cleaned up series of commit that fixes some of the issues I was seeing such as double entries showing up (being caused by a race condition). This PR allows the user to delete multiple items on Top Sites at the same time and rotate without the app crashing or something some weird state. It's kinda neat to see actually. Known issues:

1. This was happening before but the favicons have a delay and are not always being set in the correct cell. I feel like there's a race going on but I'll resolve that in a different branch.

rnewman commented:
For the record: @sleroux is looking into GCD semaphores for this.

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -110,42 +115,19 @@ class TopSitesPanel: UIViewController {
-            }
+    private func updateAllRemoveButtonStates() {
+        collection?.indexPathsForVisibleItems().forEach { indexPath in
+            updateRemoveButtonStateForIndexPath(indexPath)
```
```
collection?.indexPathsForVisibleItems()
           .forEach(updateRemoveButtonStateForIndexPath)
```

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -110,42 +115,19 @@ class TopSitesPanel: UIViewController {
-        }
+    private func updateRemoveButtonStateForIndexPath(indexPath: NSIndexPath, forCell cell: ThumbnailCell? = nil) {
+        // If we have a cell passed in, use it. If not, then use the indexPath to get it.
+        let cell = cell ?? (collection?.cellForItemAtIndexPath(indexPath) as? ThumbnailCell)
```
`guard let` here — that `as?` can still fail.

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -160,48 +142,33 @@ class TopSitesPanel: UIViewController {
-            return
+    private func invalidateTopSitesCursor() -> Success {
+        return profile.history.getTopSitesWithLimit(maxFrecencyLimit).bind { result in
+            self.cachedCursor = result.successValue
```
You should use `bindQueue` here so you always set the cached cursor on the main thread.

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -263,24 +228,57 @@ extension TopSitesPanel: UICollectionViewDelegate {
+            return
+        }
+
+        if !refillingTiles {
```
I'm trying to think of when this will be true. When they exit edit mode, immediately re-enter it and remove another tile?

sleroux replied:
`refillingTiles` gets set to true after deleting a tile and set back to false whenever `refillDeletedTiles()` finishes [1]. Refreshing the cache takes time and the user could try deleting a tile while we're refilling it. This ensures that won't happen. If we delete a tile while we're still refilling we end up having duplicates and the data source gets out of sync :(

https://github.com/mozilla/firefox-ios/pull/1283/files#diff-abfc8d88f8fbb974154bce59e22c81a6R164

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -263,24 +228,57 @@ extension TopSitesPanel: UICollectionViewDelegate {
 
 extension TopSitesPanel: ThumbnailCellDelegate {
+
     func didRemoveThumbnail(thumbnailCell: ThumbnailCell) {
```
Comment that this is always called on the main thread.

rnewman replied:
And add an assert for it!

rnewman commented on Client/Frontend/Home/TopSitesPanel.swift:
```diff
@@ -331,12 +332,19 @@ private class TopSitesLayout: UICollectionViewLayout {
-    private var width: CGFloat { return self.collectionView?.frame.width ?? 0 }
+
+    private var width: CGFloat {
+        assert(NSThread.isMainThread(), "Interacts with UIKit components - not thread-safe.")
```
Time for an `assertIsMainThread` helper?

sleroux commented:
Closing for https://github.com/mozilla/firefox-ios/pull/1394
